lib/Doctrine/DBAL/Connection.php

```diff
@@ -881,16 +880,14 @@ public function prepare($statement)
      * If the query is parametrized, a prepared statement is used.
      * If an SQLLogger is configured, the execution is logged.
      *
-     * @param string                                      $query  The SQL query to execute.
-     * @param array                                       $params The parameters to bind to the query, if any.
-     * @param array                                       $types  The types the previous parameters are in.
-     * @param \Doctrine\DBAL\Cache\QueryCacheProfile|null $qcp    The query cache profile, optional.
-     *
-     * @return \Doctrine\DBAL\Driver\Statement The executed statement.
+     * @param string                 $query  The SQL query to execute.
+     * @param mixed[]                $params The parameters to bind to the query, if any.
+     * @param mixed[]                $types  The types the previous parameters are in.
```

Comment:
Types should be `(int|string)[]`

Reply:
When iterating over `(int|string)[]`, PhpStorm will recognize elements as `mixed`, but if it's `int[]|string[]` (not exactly correct), it will recognize them as `int|string`. This was previously brought up in https://github.com/doctrine/dbal/pull/3025#discussion_r168909829.

I'm still in favor of the proper annotation instead of making PhpStorm happy, especially if it's about scalar types (@Majkl578).

Reply:
Please let's not introduce types like `(int|string)[]`. Although unfortunate, it doesn't work in PhpStorm, our CS and most of other tools either.

Reply:
Hmm I see you actually did the change... :/
I am strongly against and it should be reverted given the current state of PHP ecosystem.

Reply:
We need to achieve some agreement before reverting it. The state of the PHP ecosystem is PHP 7.1 = ~37%. Yet, we don't support the older versions.

Reply:
The state of PHP ecosystem is: PHPStan supports this, any other software doesn't.